CFE-3953: added user input support for cfbs

larsewi opened the pull request:
Here are some example modules to test with [cfbs-test-modules](https://github.com/larsewi/cfbs-test-modules)
 - [x] Write documentation
 - [x] Implement `cfbs input` command
 - [x] Remove input data on `cfbs remove`
 - [x] Prompt for input on `cfbs add`
 - [x] Update input on `cfbs update`
 - [x] Create augments on `cfbs build`

cf-bottom commented:
Thanks for submitting a pull request! Maybe @olehermanse can review this?

larsewi committed:
Moved Result namedtuple into its own file

Moved Result namedtuple into its own file in order to avoid circular
dependencies.

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
`cfbs add` now prints multiline commit messages

Ticket: None
Changelog: Title
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Added canonify utility function

This function is inspired by the `string_lib.c:CanonifyNameInPlace`
function from libntech:

```C
void CanonifyNameInPlace(char *s)
{
    for (; *s != '\0'; s++)
    {
        if (!isalnum((unsigned char) *s))
        {
            *s = '_';
        }
    }
}
```

I tested it with the CFEngine `canonify` function to see if it produces
the same results. But it responds a little bit differently when using
non-ASCII characters. E.g. the emoji 🐿️ is translated to `___` by
CFEngine but `_` by this python function.

One thing to note is that the `isalnum` function responds differently
based on what locale is set on the system.

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

vpodzime commented on cfbs/cfbs_config.py:
```diff
@@ -298,9 +299,15 @@ def add_command(
+        count = 0
+        for module in self["build"]:
+            if not any(module["name"] == m["name"] for m in before):
+                msg += "\n - Added module '%s'" % module["name"]
```
I'd prefer a set difference to be used here, but YMMV.

larsewi replied:
Like this?
```python
for name in {x["name"] for x in self["build"]}.difference(x["name"] for x in before):
    msg += "\n - Added module '%s'" % name
    count += 1
```
I think the current one is easier to read.

The following does not work:
```python
for module in set(self["build"]).difference(set(before)):
    msg += "\n - Added module '%s'" % module["name"]
    count += 1
```
```
Traceback (most recent call last):
...
  File "/usr/local/lib/python3.8/dist-packages/cfbs/cfbs_config.py", line 306, in add_command
    for module in set(self["build"]).difference(set(before)):
TypeError: unhashable type: 'collections.OrderedDict'
```

vpodzime replied:
I find the (working) set difference easiest to read only I would actually assign the two sets to variables like `module_names` and `orig_module_names`.

larsewi replied:
I agree, thanks :rocket:

vpodzime commented on cfbs/cfbs_config.py:
```diff
@@ -98,6 +99,39 @@ def save(self):
     def longest_module_name(self) -> int:
         return max((len(m["name"]) for m in self["build"]))
 
+    def _handle_input(self, input):
```
I would actually move this into `cfbs.prompts` and drop the first commit because it would eliminate the circular imports.

vpodzime replied:
Oh, I lost track of the changes in those files! I thought `CFBSConfig()` was a config not something that actually even adds and removes modules etc. Anyway, I still think this input handling should be separated. If for nothing else then at least for better reusability.

vpodzime reviewed (commented):
Looks promising otherwise. Next is handling inputs on module updates and making sure everything works with dependencies (both when adding modules and updating them).

larsewi committed:
Add additional files to commit with Result namedtuple

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
`cfbs remove` now removes input data

`cfbs remove` now removes input data. `cfbs remove` now also writes
multiline commit messages as default.

Ticket: CFE-3953
Changelog: Title
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Added docs for new cfbs input behaviour

Ticket: CFE-3953
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
`read_json` now handles `NotADirectoryError` exception

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Removed `CFBSConfig` import from prompts.py

Removed `CFBSConfig` import from prompts.py so that `prompt_user` can be
used in CFBSConfig.py.

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Improved code in `add_command` function

Changed to use set difference instead off nested for loops.

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Added command `cfbs input`

Ticket: CFE-3953
Changelog: Title
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
`cfbs add` now prompts for input

`cfbs add` now prompts users whether or not they want to add input for
modules that accept input.

Ticket: CFE-3953
Changelog: Body
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
`cfbs update` now supports modules with input

Ticket: CFE-3953
Changelog: Title
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Added .vim/ to .gitignore

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

vpodzime commented on cfbs/commands.py:
```diff
@@ -762,3 +781,87 @@ def info_command(modules):
+        # TODO: implement choices
+        string = input["string"]
+        _check_keys(("question", "label"), string)
+        return prompt_user(string["question"], default=string.get("default"))
```
`string` is not a good variable name, AFAICT.

vpodzime commented on cfbs/commands.py:
```diff
@@ -762,3 +781,87 @@ def info_command(modules):
+    def_json = {"variables": {}}
+
+    for input in module["input"]:
+        _check_keys((), input)
```
?

vpodzime commented on cfbs/commands.py:
```diff
@@ -762,3 +781,87 @@ def info_command(modules):
+
+    for input in module["input"]:
+        _check_keys((), input)
+        namespace = input["namespace"] if "namespace" in input else "cfbs"
```
`input.get("namespace", "cfbs")`

vpodzime commented on cfbs/commands.py:
```diff
@@ -803,6 +804,46 @@ def _check_keys(keys, input):
+
+    for module in build:
+        def_json = read_json(os.path.join(module["name"], "def.json"))
+        if "input" not in module or not def_json or "variables" not in def_json:
```
`def_json is None`

vpodzime commented on cfbs/cfbs_config.py:
```diff
@@ -1,4 +1,5 @@
 #!/usr/bin/env python3
+from inspect import ismethoddescriptor
```
Looks like an unused import.

vpodzime commented on cfbs/cfbs_config.py:
```diff
@@ -311,15 +311,26 @@ def add_command(
+                % name,
+                YES_NO_CHOICES,
+                "no",
+            ) in ("yes", "y"):
```
I believe the input from user needs `input.lower()`.

vpodzime commented on cfbs/commands.py:
```diff
@@ -534,24 +517,55 @@ def update_command(to_update):
                             "Please make sure the old build steps work"
                             + " with the new version of the module"
                         )
+            if key == "input":
```
`else if`

vpodzime commented on cfbs/commands.py:
```diff
@@ -534,24 +517,55 @@ def update_command(to_update):
+                if commit_differs:
+                    ans = prompt_user(
+                        config.non_interactive,
+                        "Module '%s' takes different input now. " % update.name
```
How do you know that the inputs are different?

vpodzime commented on cfbs/commands.py:
```diff
@@ -534,24 +517,55 @@ def update_command(to_update):
+                            "Do you want to add input for module '%s' right away?",
+                            choices=YES_NO_CHOICES,
+                            default="no",
+                        )
```
This second prompt is weird. The user just told us they wanted to use new inputs. I think the two prompts should be combined into a single one giving the user an option to use the defaults.

vpodzime reviewed (commented):
+ *instead of~f~`* in one of the commit messages

Looks generally OK to me. I don't like the idea of how this whole thing works, but that's another story.

larsewi commented on cfbs/result.py:
```diff
@@ -0,0 +1,12 @@
+        self.commit_msg = commit_msg
+
+    @staticmethod
+    def unpack(result) -> tuple:
```
Did not know you could do default values in constructor of namedtuples. Now I know. Thanks :beers:

larsewi replied:
> Changed in version 3.7: Added the defaults parameter and the _field_defaults attribute.

Apparently the defaults parameter was added in 3.7, thus we cannot use it. Any thoughts on how we could do this?

vpodzime replied:
You can steal from me: https://github.com/storaged-project/blivet/blob/master/blivet/util.py#L1019

vpodzime commented on JSON.md:
```diff
@@ -172,3 +172,390 @@ cfbs init && cfbs add https://github.com/cfengine/some-repo
+### Create single file example:
+The `"input"` attribute takes a list of input definitions as illustrated below.
+The value of the `"input"` attribute may be expressed as inline JSON, or as a
+file path or URL to an external JSON file.
```
Do we have a real reason/use-case for inputs specified in an external JSON file? If not, let's not complicate our lives by allowing this, at least not for now.

vpodzime commented on JSON.md:
````diff
@@ -172,3 +172,390 @@ cfbs init && cfbs add https://github.com/cfengine/some-repo
+```
+
+### Create a single file with content example:
+Modules that creates empty files is not too impressive on its own. Let us
````
*A module...*

larsewi commented on JSON.md:
```diff
@@ -172,3 +172,388 @@ cfbs init && cfbs add https://github.com/cfengine/some-repo
+class generated by the autorun dependency. Studying our variable closer, we can
+see that a namespace, bundle, and a comment, were automatically added with some
+default values.  I.e.  `cfbs`, the module name canonified, and `Added by 'cfbs
+input'` respectivy.  These defaults can easily be overrided using the
```
`s/overrided/overridden`

larsewi commented on JSON.md:
````diff
@@ -172,3 +172,388 @@ cfbs init && cfbs add https://github.com/cfengine/some-repo
+  }
+}
+```
+
````
Show that a buildstep is added in `cfbs.json` after build to make this more clear
```json
"steps": [
        "copy ./create-multiple-files-with-content.cf services/autorun/create-multiple-files-with-content.cf",
        "input ./input.json def.json"
      ],
```

larsewi commented on cfbs/commands.py:
```diff
@@ -457,88 +462,158 @@ def _someone_needs_me(this) -> bool:
     return 0
 
 
+def _update_input(module):
```
Add unit test

larsewi commented on cfbs/cfbs_config.py:
```diff
@@ -307,15 +307,31 @@ def add_command(
-            msg = msg[4:]  # Remove the '\n - ' part of the message
+
+            module = self.get_module_from_build(name)
+            if (
```
`cfbs add` should check if input needs update

larsewi committed:
Added `cfbs update` (with input) acceptance tests

Ticket: CFE-3953
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Add `cfbs update` (with input) unit tests

Ticket: CFE-3953
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Created `merge_json` unit test

Created `merge_json` unit test to check my sanity while implementing the
'input' buildstep.

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Added new 'input' build step for cfbs

`cfbs build` now generates augments when it sees the 'input' in build
steps.  If there is no input associated with the 'input' build step, no
operation is done.

Ticket: CFE-3953
Changelog: Title
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Warning when adding a module that already has input

Print a warning when adding a module that already has input data in the
module repository. In the future handle this properly. See ticket
[CFE-4059](https://tracker.mender.io/browse/CFE-4059).

Ticket: CFE-3953
Changelog: Print warning when adding a module that already has input data
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Add shell test for input.

Ticket: CFE-3953
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

vpodzime commented on cfbs/cfbs_config.py:
```diff
@@ -314,3 +316,82 @@ def add_command(
+                                % element["type"]
+                            )
+                        result[element["key"]] = _input_string(element)
+                    return result
```
This nested function somewhere inside another function is weird. I'd move it above the `_input_list()` function with one argument for the `subtype`.

vpodzime commented on cfbs/commands.py:
```diff
@@ -799,3 +800,37 @@ def info_command(modules):
+            print("Skipping module '%s', no input needed" % module_name)
+            continue
+
+        input_path = "./%s/input.json" % module_name
```
`os.path.join()`

vpodzime reviewed (commented):
I've only gotten to *Added command cfbs input*, will follow up with the rest of the commits later. Apart from the nitpicks below, please make sure that the docs added in the second commit still actually reflect what the commits after it introduce (and maybe move it to the end of the commits in this PR?). Thanks for working on this!

vpodzime commented on cfbs/commands.py:
```diff
@@ -453,88 +458,181 @@ def _someone_needs_me(this) -> bool:
+        input_data (dict): Input data to update
+
+    Returns:
+        bool: True if changes are made
```
Might as well use the _Sphinx_ syntax here (see the docstrings in *git.py*).

vpodzime commented on cfbs/commands.py:
```diff
@@ -49,6 +50,10 @@
 from cfbs.module import Module
 
 
+class InputDataUpdateFailedException(Exception):
```
I think the `Exception` part is not needed in the name of the exception class. If you want to have something there, `Error` is usually used.

vpodzime commented on cfbs/commands.py:
```diff
@@ -453,88 +458,181 @@ def _someone_needs_me(this) -> bool:
+            + "Input definition has %d variables, " % len(input_def)
+            + "while current input data has %d variables." % len(input_data)
+        )
+        raise InputDataUpdateFailedException()
```
I think the exception should also contain the message/reason).

vpodzime commented on cfbs/commands.py:
```diff
@@ -453,88 +458,181 @@ def _someone_needs_me(this) -> bool:
+        [Module(m) for m in to_update]
+        if to_update
+        else [Module(m["name"]) for m in build]
+    )
```
This part (and many changes below) doesn't seem to be related to handling inputs as part of `update`. Ideally, it should be in a separate commit, but not a big deal if it would be hard to separate.

vpodzime commented on tests/shell/all.sh:
```diff
@@ -20,3 +20,7 @@ bash tests/shell/014_add_nonexistent.sh
+bash tests/shell/018_update_input_one_variable.sh
+bash tests/shell/019_update_input_two_variables.sh
+bash tests/shell/020_update_input_list.sh
+bash tests/shell/021_update_input_list_with_keys.sh
```
Would be nice to have tests for the bad cases too.

vpodzime commented on cfbs/build.py:
```diff
@@ -1,5 +1,7 @@
 import os
+import logging as log
```
*If input associated with the 'input' build step, no operation is done.* -- missing *no* in the commit message

vpodzime commented on cfbs/build.py:
```diff
@@ -106,6 +144,36 @@ def _perform_build_step(module, step, max_length):
+            log.warning(
+                "Input data '%s' is incomplete: Skipping build step."
+                % os.path.basename(src)
+            )
```
This should be an error.

vpodzime commented on cfbs/commands.py:
```diff
@@ -382,9 +382,8 @@ def _get_modules_by_url(name) -> list:
-                name, input_path
-            ),
+            "Module '%s' has input data '%s'. Do you want to remove it?"
+            % (name, input_path),
```
This fix should go to a different commit.

vpodzime commented on cfbs/cfbs_config.py:
```diff
@@ -316,23 +316,26 @@ def add_command(
+                    log.warning(
+                        "There seems to already be input for module '%s'. " % name
+                        + "Note that old input may break the module. "
+                        + "Please make sure to run 'cfbs update', "
```
This is confusing. I'd expect `cfbs input MODULE` to be the command to run.

vpodzime commented on tests/shell/022_add_input_remove.sh:
```diff
@@ -0,0 +1,14 @@
+
+cfbs --non-interactive add example-module
+cfbs --non-interactive input example-module
+cat ./example-module/input.json | grep '"response": "/tmp/testfile.txt"'
```
`grep` takes a file path argument ;)

vpodzime reviewed (commented):
Suggestions for the rest of the commits.

larsewi committed:
`cfbs update` now supports custom index

Ticket: None
Changelog: Title
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Change regex patterns to be raw strings

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Print message when module is already up to date

Print message when trying to update a module that is already up to date.

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Change log message to use '%' string formatting

My linter complained and said this was better. I don't know why, but I
trust my linter.

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Fixed bug where updating build step does not commit

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi commented on cfbs/build.py:
```diff
@@ -21,6 +23,42 @@ def init_out_folder():
 
 
+def _generate_augment(module_name, input_data):
+    """
+    Generate augment from input data.
+
+    Params:
+        module_name (str): Name of module
+        input_data (list): Input data
+
+    Returns:
+        dict: Generated augment or None if input data is incomplete
+    """
```
Change to Sphinx syntax here as well

larsewi commented on tests/shell/021_update_input_list_with_keys.sh:
```diff
@@ -0,0 +1,22 @@
+cp ../shell/021_update_input_list_with_keys/example-cfbs.json cfbs.json
+
+cfbs --loglevel=debug --non-interactive update
+cat example-module/input.json | grep '"while": "Create another file?"'
+cat example-module/input.json | grep '"label": "Path"'
+cat example-module/input.json | grep '"question": "Path of file?"'
+cat example-module/input.json | grep '"default": "/tmp/test.txt"'
+cat example-module/input.json | grep '"label": "Contents"'
+cat example-module/input.json | grep '"question": "File contents?"'
+cat example-module/input.json | grep '"default": "Hello CFEngine!"'
+cat example-module/input.json | grep '"response":'
+cat example-module/input.json | grep '{ "name": "/tmp/test-1.txt", "content": "Hello CFEngine!" }'
+cat example-module/input.json | grep '{ "name": "/tmp/test-2.txt", "content": "Bye CFEngine!" }'
```
Use grep argument here as well

larsewi commented on tests/shell/020_update_input_list.sh:
```diff
@@ -0,0 +1,20 @@
+cp ../shell/020_update_input_list/example-cfbs.json cfbs.json
+
+cfbs --loglevel=debug --non-interactive update
+cat example-module/input.json | grep '"label": "Filepaths"'
+cat example-module/input.json | grep '"while": "Create another file?"'
+cat example-module/input.json | grep '"label": "Path"'
+cat example-module/input.json | grep '"question": "Path of file?"'
+cat example-module/input.json | grep '"default": "/tmp/test.txt"'
+cat example-module/input.json | grep '"response":'
+cat example-module/input.json | grep '"/tmp/create-multiple-files-1.txt"'
+cat example-module/input.json | grep '"/tmp/create-multiple-files-2.txt"'
```
Use grep argument here as well

larsewi commented on tests/shell/019_update_input_two_variables.sh:
```diff
@@ -0,0 +1,19 @@
+cp ../shell/019_update_input_two_variables/example-cfbs.json cfbs.json
+
+cfbs --loglevel=debug --non-interactive update
+cat example-module/input.json | grep '"label": "Path"'
+cat example-module/input.json | grep '"question": "Path of file?"'
+cat example-module/input.json | grep '"default": "/tmp/test.txt"'
+cat example-module/input.json | grep '"label": "Contents"'
+cat example-module/input.json | grep '"question": "File contents?"'
+cat example-module/input.json | grep '"default": "Hello CFEngine!"'
+cat example-module/input.json | grep '"response": "/tmp/create-single-file-with-content.txt"'
```
Use grep argument here as well

olehermanse commented on cfbs/utils.py:
```diff
@@ -281,3 +281,7 @@ def wrapper(*args, **kwargs):
+
+
+def canonify(string):
+    return re.sub(r"[^a-zA-Z0-9_\200-\377]", "_", string)
```
Should probably reference where this is from (in commit message).

olehermanse commented on tests/shell/025_add_input_remove.sh:
```diff
@@ -0,0 +1,14 @@
+cfbs --non-interactive add example-module
+cfbs --non-interactive input example-module
+grep '"response": "/tmp/testfile.txt"' ./example-module/input.json
+cfbs --non-interactive remove example-module
```
Should probably check here what remove did.

olehermanse reviewed (approved):
Looks good to me, thanks for working on this!
